### ISSUE TITLE
Handle outdated versions of TGMPA

### ISSUE THREAD

**Issue by mrazzari** [itspriddle/wp-cli-tgmpa-plugin#21]
So I randomly came across an old-ish version of this [iribbon](https://wordpress.org/themes/iribbon/) theme, and here's what happens:

```
$ wp theme install https://downloads.wordpress.org/theme/iribbon.2.0.65.zip --activate
$ wp tgmpa-plugin list
PHP Fatal error:  Uncaught Error: Call to undefined method TGM_Plugin_Activation::get_instance() 
in /home/manuel/.wp-cli/packages/vendor/itspriddle/wp-cli-tgmpa-plugin/command.php:70
```

Turns out the theme is using TGMPA 2.3.6, which doesn't have a get_instance (added in 2.4.0). 

You could catch that exception, and use another way to get the instance. But it could start breaking in some other places. So maybe just throw a nice WP_CLI::error and declare it unsupported? :roll_eyes:

**Comment by itspriddle** [comment]
Let me play around a bit and see if I can come up with something that works well without introducing a lot of complexity.

**Comment by itspriddle** [comment]
First, on required TGMPA versions: I don't really have enough real world usage to feel comfortable requiring specific TGMPA versions yet. For now I plan to support any version that I reasonably can. If, for example, supporting TGMPA 2.0.0 required a ton of work, that would likely be where I'd draw the line. But I think a few methods here and there to work around behavioral differences between versions are worth the maintenance cost to make the project usable with a wider selection of themes/plugins.

Looking into iribbon there are multiple issues at play. The first is the one in your original post, that `TGM_Plugin_Activation::get_instance()` is not defined. As far as I can tell, in TGMPA 2.3.6 it is safe to use `TGM_Plugin_Activation::$instance` as an alternative as long as `do_action("tgmpa_register")` has been called first. I've tweaked `wp tgmpa-plugin` to fallback if necessary in #22.

The second issue is that the `tgmpa_register` action is not set when `wp tgmpa-plugin` loads -- which is a bit of a rabbit hole as to why. iribbon checks `current_user_can("manage_options")` or it does not set `tgmpa_register`. Normally, you can just `wp --user=admin` with a valid admin user to have such a check pass. But iribbon attempts to set `tgmpa_register` inside of an `after_setup_theme` hook, and this hook fires before the `init` hook that WP-CLI uses to assign a current user, so the check fails and the `tgmpa_register` action is not defined. 

So what to do?

`wp tgmpa-plugin` should definitely check for the `tgmpa_register` action and show a proper error message if it is missing. I've added that in #22.

I suspect there's probably a Good Reason™ WP-CLI uses `init` when it assigns a current user. I'm a bit hesitant to add something to `wp tgmpa-plugin` to override that behavior, especially until I've seen more cases I can test. To be clear, I'm not at all against doing this, but I want to make sure it is really necessary first and can be done in a way that supports as many setups as possible.

For now, if you need a way to install iribbon with WP-CLI (after #22 ships), you can use a custom require file similar to this:

```php
<?php // iribbon-fix.php

function fix_iribbon() {
  // Replace this with whatever logic you want for setting the current user
  wp_set_current_user(1);
}

WP_CLI::add_wp_hook("after_setup_theme", "fix_iribbon");
```

Then you can invoke the CLI like so:

```
wp --require=iribbon-fix.php tgmpa-plugin ...
```

That should fix the issue by assigning a user earlier in the request so that iribbon's permission check will pass.

Let me know if you have any other thoughts, and thanks for taking the time to open the bug report!

**Comment by itspriddle** [comment]
Closed by #22, #23

**Comment by mrazzari** [comment]
That's great to hear, thanks for the fix! 
Shocked to hear you're willing to try and support older versions of TGMPA.  I wouldn't dare ;)

And thanks for the iribbon-fix idea. I don't have a use for it right now, but it's a great hack that will surely come useful.
